src/utils/clientHelpers.ts

```diff
@@ -1,18 +1,56 @@
-import { Executable, InitializeResult } from 'vscode-languageclient/node';
+import * as net from 'net';
+import * as vscode from 'vscode';
+import { Executable, InitializeResult, ServerOptions, StreamInfo } from 'vscode-languageclient/node';
 import { config } from './vscode';
 import { ServerPath } from './serverPath';
+import { outputChannel } from '../extension';
+
+export async function getServerOptions(lsPath: ServerPath): Promise<ServerOptions> {
+  let serverOptions: ServerOptions;
+
+  const port: number | undefined = config('terraform').get('experimentalFeatures.languageServer.tcp.port');
+  if (port) {
+    const inspect = vscode.workspace.getConfiguration('terraform').inspect('languageServer.path');
+    if (inspect !== undefined) {
+      ///
+      if (inspect.globalValue || inspect.workspaceFolderValue || inspect.workspaceValue) {
+        vscode.window.showWarningMessage(
+          'You cannot use experimentalFeatures.languageServer.tcp.port with terraform.languageServer.path. Ignoring terraform.languageServer.path and proceeding to connect via TCP',
+        );
+      }
+    }
+
+    serverOptions = () => {
```

Comment:
```suggestion
    serverOptions = async () => {
```

If we make this function `async` we can avoid the manual creation of a promise. We could even return a bit earlier:
```ts
return {
  writer: socket,
  reader: socket,
};
```